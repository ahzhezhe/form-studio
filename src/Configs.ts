import { ChoiceOnSelected, ChoiceValue, CustomConfigs, QuestionType } from './Types';

/**
 * @category Configs
 */
export interface ItemConfigs {
  /**
   * Item id.
   * It should be unique throughout the entire form.
   * It will be auto generated if you leave it falsy.
   */
  id?: string;
  /**
   * Item sort order.
   * Items will be sorted by their positions in the list if their order is `undefined`.
   */
  order?: number;
  /**
   * Whether or not this item is disabled by default.
   * The actual disabled property will be based on the logic that you define in other part of the configs.
   */
  defaultDisabled?: boolean;
  /**
   * Any values that help you determine on how to render the frontend UI for this item.
   */
  ui?: CustomConfigs;
}

/**
 * @category Configs
 */
export interface GroupConfigs extends ItemConfigs {
  /**
   * Sub-groups.
   */
  groups?: GroupConfigs[];
  /**
   * Questions under this group.
   */
  questions?: QuestionConfigs[];
}

/**
 * @category Configs
 */
export interface QuestionConfigs extends ItemConfigs {
  /**
   * Question type.
   *
   * Questions with `any` as [[type]] accept `any` as answer.
   * You shouldn't have to specify [[choices]] for this question.
   *
   * Questions with `single` as [[type]] accept [[ChoiceValue]] as answer.
   * You should specify [[choices]] for this question.
   * Usually radio button group will be used for this question.
   *
   * Questions with `multiple` as [[type]] accept [[ChoiceValue]][] as answer.
   * You should specify [[choices]] for this question.
   * Usually check box group will be used for this question.
   */
  type: QuestionType;
  /**
   * Choices for questions with `single` or `multiple` as [[type]].
   */
  choices?: ChoiceConfigs[];
  /**
   * Names of the validators to be used for validation when answer of this question is changed.
   *
   * Validators will be executed in sequence of their positions in the list.
   */
  validators?: string[];
  /**
   * Any values that help to validator to perform validation.
   */
  validation?: CustomConfigs;
  /**
   * Default answer.
   *
   * Questions with `any` as [[type]] accept `any` as answer.
   *
   * Questions with `single` as [[type]] accept [[ChoiceValue]] as answer.
   *
   * Questions with `multiple` as [[type]] accept [[ChoiceValue]][] as answer.
   */
  defaultAnswer?: any;
}

/**
 * @category Configs
 */
export interface ChoiceConfigs extends ItemConfigs {
  /**
   * Value of this choice. It will be used as answer of the question.
   * Id will be used as value if you leave it `undefined`.
   */
  value?: ChoiceValue;
  /**
   * Actions to be performed when this choice is toggled.
   */
  onSelected?: ChoiceOnSelected;
}

/**
 * Configs to construct a form.
 *
 * @category Configs
 */
export type Configs = GroupConfigs[];
