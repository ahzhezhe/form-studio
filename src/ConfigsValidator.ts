import { ExportedConfigs, Group, Question } from './ExportedConfigs';
import { ConfigsValidationResult } from './Types';

export class ConfigsValidator {

  readonly #errorsById = new Map<string, string[]>();
  readonly #allIds: string[] = [];
  readonly #choiceValuesByQuestionId = new Map<string, any[]>();
  readonly #onSelectedIdsByChoiceId = new Map<string, string[]>();
  readonly #parentIdsById = new Map<string, string[]>();
  readonly #circularIds: string[] = [];

  validate(configs: ExportedConfigs, strict: boolean): ConfigsValidationResult {
    // Form is not without any groups or questions
    if (configs.groups.length === 0 && configs.questions.length === 0) {
      this.#addError('', 'There are no groups or questions');
      return this.#getResult();
    }

    // Collect data for subsequent validations
    this.#collectData(configs);

    // No duplicated ids within the form
    const duplicatedIds = this.#findDuplicates(this.#allIds);
    duplicatedIds.forEach(id => this.#addError(id, 'Id is not unique'));

    // No duplicated choice values within a question
    for (const [questionId, values] of this.#choiceValuesByQuestionId.entries()) {
      const duplicatedValues = this.#findDuplicates(values);
      if (duplicatedValues.length) {
        this.#addError(questionId, 'There are choices with same values');
      }
    }

    // No circular choices' `onSelected` configs
    if (!duplicatedIds.length) {
      for (const id of this.#allIds) {
        this.#checkCircular(id, id);
      }
    }

    if (strict) {
      // No unrecognized ids in choices' `onSelected` configs
      for (const [choiceId, onSelectedIds] of this.#onSelectedIdsByChoiceId.entries()) {
        for (const onSelectedId of onSelectedIds) {
          if (!this.#allIds.includes(onSelectedId)) {
            this.#addError(choiceId, 'There are unrecognized id(s) in onSelected configs');
            break;
          }
        }
      }
    }

    return this.#getResult();
  }

  #checkCircular(mainId: string, subId: string) {
    const parentIds = this.#parentIdsById.get(subId)!;

    if (parentIds.includes(mainId)) {
      this.#addError(mainId, `Circular relationship with '${subId}'`);
      this.#circularIds.push(mainId);
      return;
    }

    for (const parentId of parentIds) {
      if (this.#circularIds.includes(parentId)) {
        continue;
      }
      this.#checkCircular(mainId, parentId);
    }
  }

  #collectData(configs: ExportedConfigs) {
    this.#collectGroupData(undefined, configs.groups);

    for (const [choiceId, onSelectedIds] of this.#onSelectedIdsByChoiceId) {
      for (const onSelectedId of onSelectedIds) {
        const parentIds = this.#parentIdsById.get(onSelectedId);
        if (parentIds) {
          parentIds.push(choiceId);
          this.#parentIdsById.set(onSelectedId, parentIds);
        }
      }
    }
  }

  #collectGroupData(parentGroupId: string | undefined, groups: Group[]) {
    for (const group of groups) {
      this.#allIds.push(group.id);
      if (parentGroupId) {
        this.#parentIdsById.set(group.id, [parentGroupId]);
      } else {
        this.#parentIdsById.set(group.id, []);
      }
      this.#collectGroupData(group.id, group.groups);

      // No groups without questions
      if (group.questions.length === 0) {
        this.#addError(group.id, 'There are no questions');
        continue;
      }

      this.#collectQuestionData(group.id, group.questions);
    }
  }

  #collectQuestionData(groupId: string | undefined, questions: Question[]) {
    for (const question of questions) {
      this.#allIds.push(question.id);
      if (groupId) {
        this.#parentIdsById.set(question.id, [groupId]);
      } else {
        this.#parentIdsById.set(question.id, []);
      }

      // No questions with `choice` or `choices` as type without choices
      if (question.type !== 'any') {
        if (question.choices.length === 0) {
          this.#addError(groupId || '', 'There are no choices');
          continue;
        }

        for (const choice of question.choices) {
          this.#allIds.push(choice.id);
          this.#parentIdsById.set(choice.id, [question.id]);
          const onSelectedIds = [...(choice.onSelected.enable || []), ...(choice.onSelected.disable || [])];
          if (onSelectedIds.length) {
            this.#onSelectedIdsByChoiceId.set(choice.id, onSelectedIds);
          }
        }
        this.#choiceValuesByQuestionId.set(question.id, question.choices.map(choice => choice.value));
      }
    }
  }

  #findDuplicates <T>(arr: T[]) {
    const sorted = arr.slice().sort();
    const duplicatedIds = new Set<T>();
    for (let i = 0; i < sorted.length - 1; i++) {
      if (sorted[i + 1] === sorted[i]) {
        duplicatedIds.add(sorted[i]);
      }
    }
    return Array.from(duplicatedIds);
  }

  #addError(id: string, error: string) {
    let errors = this.#errorsById.get(id);
    if (!errors) {
      errors = [];
    }
    errors.push(error);
    this.#errorsById.set(id, errors);
  }

  #getResult(): ConfigsValidationResult {
    if (this.#errorsById.size) {
      const errors: Record<string, string[]> = {};
      this.#errorsById.forEach((values, key) => {
        errors[key] = values;
      });
      return { valid: false, errors };
    }
    return { valid: true };
  }

}
