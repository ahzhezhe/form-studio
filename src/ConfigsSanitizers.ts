import { ChoiceConfigs, Configs, GroupConfigs, QuestionConfigs } from './Configs';
import { Choice, ExportedConfigs, Group, Question } from './ExportedConfigs';

export const sanitizeConfigs = (configs: Configs): ExportedConfigs => ({
  groups: configs.groups ? sanitizeGroupConfigs(undefined, configs.groups) : [],
  questions: configs.questions ? sanitizeQuestionConfigs(undefined, configs.questions) : []
});

const sanitizeGroupConfigs = (parentGroupId: string | undefined, groups: GroupConfigs[]): Group[] => groups.map((group, i) => {
  const id = group.id || (parentGroupId ? `${parentGroupId}_g${i}` : `g${i}`);
  return {
    id,
    defaultDisabled: !!group.defaultDisabled,
    custom: group.custom,
    groups: group.groups ? sanitizeGroupConfigs(id, group.groups) : [],
    questions: group.questions ? sanitizeQuestionConfigs(id, group.questions) : []
  };
});

const sanitizeQuestionConfigs = (groupId: string | undefined, questions: QuestionConfigs[]): Question[] => questions.map((question, i) => {
  const id = question.id || (groupId ? `${groupId}_q${i}` : `q${i}`);
  return {
    id,
    defaultDisabled: !!question.defaultDisabled,
    custom: question.custom,
    type: question.type,
    choices: question.type !== 'any' ? sanitizeChoiceConfigs(id, question.choices || []) : [],
    validators: question.validators || [],
    defaultAnswer: question.defaultAnswer
  };
});

const sanitizeChoiceConfigs = (questionId: string, choices: ChoiceConfigs[]): Choice[] => choices.map((choice, i) => {
  const id = choice.id || `${questionId}_c${i}`;
  return {
    id,
    defaultDisabled: !!choice.defaultDisabled,
    custom: choice.custom,
    value: choice.value === undefined || choice.value === null ? id : choice.value,
    onSelected: choice.onSelected || {}
  };
});
