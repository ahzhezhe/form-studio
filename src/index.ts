/* eslint-disable import/no-default-export */
import { GroupConfig, QuestionConfig, ChoiceConfig, Config } from './Configs';
import { FormEngine } from './FormEngine';
import { GroupInitConfig, QuestionInitConfig, ChoiceInitConfig, InitConfig } from './InitConfigs';
import { GroupTemplate, QuestionTemplate, ChoiceTemplate, Template } from './Templates';
import { CustomConfig, QuestionType, ChoiceValue, ChoiceOnSelected, Validator, Answers, Errors } from './Types';
import { useTemplate } from './useTemplate';

export {
  GroupConfig, QuestionConfig, ChoiceConfig, Config,
  GroupInitConfig, QuestionInitConfig, ChoiceInitConfig, InitConfig,
  GroupTemplate, QuestionTemplate, ChoiceTemplate, Template,
  CustomConfig, QuestionType, ChoiceValue, ChoiceOnSelected, Validator, Answers, Errors,
  useTemplate
};

export default FormEngine;
