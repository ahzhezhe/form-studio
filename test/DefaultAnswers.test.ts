import { Form } from '../src/Form';
import { validators, getConfigs } from '.';

describe('Default Answers', () => {

  const form = new Form(getConfigs(true), { validators });

  test('resetAnswer', () => {
    form.resetAnswer('subGroup1Question1');

    expect(form.getRenderInstructions()).toMatchSnapshot();
    expect(form.getErrors()).toMatchSnapshot();
    expect(form.getCurrentAnswers()).toMatchSnapshot();
    expect(form.getValidatedAnswers()).toMatchSnapshot();
  });

  test('resetGroup', () => {
    form.resetGroup('subGroup1');

    expect(form.getRenderInstructions()).toMatchSnapshot();
    expect(form.getErrors()).toMatchSnapshot();
    expect(form.getCurrentAnswers()).toMatchSnapshot();
    expect(form.getValidatedAnswers()).toMatchSnapshot();
  });

  test('reset', () => {
    form.reset();

    expect(form.getRenderInstructions()).toMatchSnapshot();
    expect(form.getErrors()).toMatchSnapshot();
    expect(form.getCurrentAnswers()).toMatchSnapshot();
    expect(form.getValidatedAnswers()).toMatchSnapshot();
  });

});
