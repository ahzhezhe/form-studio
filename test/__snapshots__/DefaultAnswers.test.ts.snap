// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Default Answers reset 1`] = `
Array [
  Object {
    "disabled": false,
    "groups": Array [
      Object {
        "disabled": false,
        "groups": Array [],
        "id": "subGroup1",
        "questions": Array [
          Object {
            "choices": Array [],
            "currentAnswer": "subGroup1Question1Value",
            "disabled": false,
            "error": undefined,
            "id": "subGroup1Question1",
            "type": "any",
            "ui": Object {
              "title": "subGroup1Questions",
            },
            "validatedAnswer": "subGroup1Question1Value",
            "validating": false,
          },
          Object {
            "choices": Array [
              Object {
                "disabled": false,
                "id": "subGroup1Question2Choice1",
                "ui": Object {
                  "title": "subGroup1Question2Choice1",
                },
                "value": "subGroup1Question2Choice1",
              },
              Object {
                "disabled": false,
                "id": "subGroup1Question2Choice2",
                "ui": Object {
                  "title": "subGroup1Question2Choice2",
                },
                "value": "subGroup1Question2Choice2",
              },
              Object {
                "disabled": false,
                "id": "subGroup1Question2Choice3",
                "ui": Object {
                  "title": "subGroup1Question2Choice3",
                },
                "value": "subGroup1Question2Choice3",
              },
            ],
            "currentAnswer": "subGroup1Question2Choice1",
            "disabled": false,
            "error": undefined,
            "id": "subGroup1Question2",
            "type": "single",
            "ui": Object {
              "title": "subGroup1Question2",
            },
            "validatedAnswer": "subGroup1Question2Choice1",
            "validating": false,
          },
          Object {
            "choices": Array [
              Object {
                "disabled": false,
                "id": "subGroup1Question3Choice1",
                "ui": Object {
                  "title": "subGroup1Question3Choice1",
                },
                "value": "subGroup1Question3Choice1",
              },
              Object {
                "disabled": false,
                "id": "subGroup1Question3Choice2",
                "ui": Object {
                  "title": "subGroup1Question3Choice2",
                },
                "value": "subGroup1Question3Choice2",
              },
              Object {
                "disabled": false,
                "id": "subGroup1Question3Choice3",
                "ui": Object {
                  "title": "subGroup1Question3Choice3",
                },
                "value": "subGroup1Question3Choice3",
              },
            ],
            "currentAnswer": Array [
              "subGroup1Question3Choice1",
              "subGroup1Question3Choice2",
            ],
            "disabled": false,
            "error": undefined,
            "id": "subGroup1Question3",
            "type": "multiple",
            "ui": Object {
              "title": "subGroup1Question3",
            },
            "validatedAnswer": Array [
              "subGroup1Question3Choice1",
              "subGroup1Question3Choice2",
            ],
            "validating": false,
          },
          Object {
            "choices": Array [
              Object {
                "disabled": false,
                "id": "subGroup1Question4Choice1",
                "ui": Object {
                  "title": "subGroup1Question4Choice1",
                },
                "value": "subGroup1Question4Choice1",
              },
              Object {
                "disabled": false,
                "id": "subGroup1Question4Choice2",
                "ui": Object {
                  "title": "subGroup1Question4Choice2",
                },
                "value": "subGroup1Question4Choice2",
              },
              Object {
                "disabled": false,
                "id": "subGroup1Question4Choice3",
                "ui": Object {
                  "title": "subGroup1Question4Choice3",
                },
                "value": "subGroup1Question4Choice3",
              },
            ],
            "currentAnswer": Array [
              "subGroup1Question4Choice1",
              "subGroup1Question4Choice2",
            ],
            "disabled": false,
            "error": undefined,
            "id": "subGroup1Question4",
            "type": "multiple",
            "ui": Object {
              "title": "subGroup1Question3",
            },
            "validatedAnswer": Array [
              "subGroup1Question4Choice1",
              "subGroup1Question4Choice2",
            ],
            "validating": false,
          },
        ],
        "ui": Object {},
      },
    ],
    "id": "group1",
    "questions": Array [
      Object {
        "choices": Array [],
        "currentAnswer": "group1Question1Value",
        "disabled": false,
        "error": undefined,
        "id": "group1Question1",
        "type": "any",
        "ui": Object {
          "title": "group1Questions",
        },
        "validatedAnswer": "group1Question1Value",
        "validating": false,
      },
      Object {
        "choices": Array [
          Object {
            "disabled": false,
            "id": "group1Question2Choice1",
            "ui": Object {
              "title": "group1Question2Choice1",
            },
            "value": "group1Question2Choice1",
          },
          Object {
            "disabled": false,
            "id": "group1Question2Choice2",
            "ui": Object {
              "title": "group1Question2Choice2",
            },
            "value": "group1Question2Choice2",
          },
          Object {
            "disabled": false,
            "id": "group1Question2Choice3",
            "ui": Object {
              "title": "group1Question2Choice3",
            },
            "value": "group1Question2Choice3",
          },
        ],
        "currentAnswer": "group1Question2Choice1",
        "disabled": false,
        "error": undefined,
        "id": "group1Question2",
        "type": "single",
        "ui": Object {
          "title": "group1Question2",
        },
        "validatedAnswer": "group1Question2Choice1",
        "validating": false,
      },
      Object {
        "choices": Array [
          Object {
            "disabled": false,
            "id": "group1Question3Choice1",
            "ui": Object {
              "title": "group1Question3Choice1",
            },
            "value": "group1Question3Choice1",
          },
          Object {
            "disabled": false,
            "id": "group1Question3Choice2",
            "ui": Object {
              "title": "group1Question3Choice2",
            },
            "value": "group1Question3Choice2",
          },
          Object {
            "disabled": false,
            "id": "group1Question3Choice3",
            "ui": Object {
              "title": "group1Question3Choice3",
            },
            "value": "group1Question3Choice3",
          },
        ],
        "currentAnswer": Array [
          "group1Question3Choice1",
          "group1Question3Choice2",
        ],
        "disabled": false,
        "error": undefined,
        "id": "group1Question3",
        "type": "multiple",
        "ui": Object {
          "title": "group1Question3",
        },
        "validatedAnswer": Array [
          "group1Question3Choice1",
          "group1Question3Choice2",
        ],
        "validating": false,
      },
    ],
    "ui": Object {},
  },
]
`;

exports[`Default Answers reset 2`] = `Object {}`;

exports[`Default Answers reset 3`] = `
Object {
  "group1Question1": "group1Question1Value",
  "group1Question2": "group1Question2Choice1",
  "group1Question3": Array [
    "group1Question3Choice1",
    "group1Question3Choice2",
  ],
  "subGroup1Question1": "subGroup1Question1Value",
  "subGroup1Question2": "subGroup1Question2Choice1",
  "subGroup1Question3": Array [
    "subGroup1Question3Choice1",
    "subGroup1Question3Choice2",
  ],
  "subGroup1Question4": Array [
    "subGroup1Question4Choice1",
    "subGroup1Question4Choice2",
  ],
}
`;

exports[`Default Answers reset 4`] = `
Object {
  "group1Question1": "group1Question1Value",
  "group1Question2": "group1Question2Choice1",
  "group1Question3": Array [
    "group1Question3Choice1",
    "group1Question3Choice2",
  ],
  "subGroup1Question1": "subGroup1Question1Value",
  "subGroup1Question2": "subGroup1Question2Choice1",
  "subGroup1Question3": Array [
    "subGroup1Question3Choice1",
    "subGroup1Question3Choice2",
  ],
  "subGroup1Question4": Array [
    "subGroup1Question4Choice1",
    "subGroup1Question4Choice2",
  ],
}
`;

exports[`Default Answers resetAnswer 1`] = `
Array [
  Object {
    "disabled": false,
    "groups": Array [
      Object {
        "disabled": false,
        "groups": Array [],
        "id": "subGroup1",
        "questions": Array [
          Object {
            "choices": Array [],
            "currentAnswer": "subGroup1Question1Value",
            "disabled": false,
            "error": undefined,
            "id": "subGroup1Question1",
            "type": "any",
            "ui": Object {
              "title": "subGroup1Questions",
            },
            "validatedAnswer": "subGroup1Question1Value",
            "validating": false,
          },
          Object {
            "choices": Array [
              Object {
                "disabled": false,
                "id": "subGroup1Question2Choice1",
                "ui": Object {
                  "title": "subGroup1Question2Choice1",
                },
                "value": "subGroup1Question2Choice1",
              },
              Object {
                "disabled": false,
                "id": "subGroup1Question2Choice2",
                "ui": Object {
                  "title": "subGroup1Question2Choice2",
                },
                "value": "subGroup1Question2Choice2",
              },
              Object {
                "disabled": false,
                "id": "subGroup1Question2Choice3",
                "ui": Object {
                  "title": "subGroup1Question2Choice3",
                },
                "value": "subGroup1Question2Choice3",
              },
            ],
            "currentAnswer": "subGroup1Question2Choice1",
            "disabled": false,
            "error": undefined,
            "id": "subGroup1Question2",
            "type": "single",
            "ui": Object {
              "title": "subGroup1Question2",
            },
            "validatedAnswer": "subGroup1Question2Choice1",
            "validating": false,
          },
          Object {
            "choices": Array [
              Object {
                "disabled": false,
                "id": "subGroup1Question3Choice1",
                "ui": Object {
                  "title": "subGroup1Question3Choice1",
                },
                "value": "subGroup1Question3Choice1",
              },
              Object {
                "disabled": false,
                "id": "subGroup1Question3Choice2",
                "ui": Object {
                  "title": "subGroup1Question3Choice2",
                },
                "value": "subGroup1Question3Choice2",
              },
              Object {
                "disabled": false,
                "id": "subGroup1Question3Choice3",
                "ui": Object {
                  "title": "subGroup1Question3Choice3",
                },
                "value": "subGroup1Question3Choice3",
              },
            ],
            "currentAnswer": Array [
              "subGroup1Question3Choice1",
              "subGroup1Question3Choice2",
            ],
            "disabled": false,
            "error": undefined,
            "id": "subGroup1Question3",
            "type": "multiple",
            "ui": Object {
              "title": "subGroup1Question3",
            },
            "validatedAnswer": Array [
              "subGroup1Question3Choice1",
              "subGroup1Question3Choice2",
            ],
            "validating": false,
          },
          Object {
            "choices": Array [
              Object {
                "disabled": false,
                "id": "subGroup1Question4Choice1",
                "ui": Object {
                  "title": "subGroup1Question4Choice1",
                },
                "value": "subGroup1Question4Choice1",
              },
              Object {
                "disabled": false,
                "id": "subGroup1Question4Choice2",
                "ui": Object {
                  "title": "subGroup1Question4Choice2",
                },
                "value": "subGroup1Question4Choice2",
              },
              Object {
                "disabled": false,
                "id": "subGroup1Question4Choice3",
                "ui": Object {
                  "title": "subGroup1Question4Choice3",
                },
                "value": "subGroup1Question4Choice3",
              },
            ],
            "currentAnswer": Array [
              "subGroup1Question4Choice1",
              "subGroup1Question4Choice2",
            ],
            "disabled": false,
            "error": undefined,
            "id": "subGroup1Question4",
            "type": "multiple",
            "ui": Object {
              "title": "subGroup1Question3",
            },
            "validatedAnswer": Array [
              "subGroup1Question4Choice1",
              "subGroup1Question4Choice2",
            ],
            "validating": false,
          },
        ],
        "ui": Object {},
      },
    ],
    "id": "group1",
    "questions": Array [
      Object {
        "choices": Array [],
        "currentAnswer": "group1Question1Value",
        "disabled": false,
        "error": undefined,
        "id": "group1Question1",
        "type": "any",
        "ui": Object {
          "title": "group1Questions",
        },
        "validatedAnswer": "group1Question1Value",
        "validating": false,
      },
      Object {
        "choices": Array [
          Object {
            "disabled": false,
            "id": "group1Question2Choice1",
            "ui": Object {
              "title": "group1Question2Choice1",
            },
            "value": "group1Question2Choice1",
          },
          Object {
            "disabled": false,
            "id": "group1Question2Choice2",
            "ui": Object {
              "title": "group1Question2Choice2",
            },
            "value": "group1Question2Choice2",
          },
          Object {
            "disabled": false,
            "id": "group1Question2Choice3",
            "ui": Object {
              "title": "group1Question2Choice3",
            },
            "value": "group1Question2Choice3",
          },
        ],
        "currentAnswer": "group1Question2Choice1",
        "disabled": false,
        "error": undefined,
        "id": "group1Question2",
        "type": "single",
        "ui": Object {
          "title": "group1Question2",
        },
        "validatedAnswer": "group1Question2Choice1",
        "validating": false,
      },
      Object {
        "choices": Array [
          Object {
            "disabled": false,
            "id": "group1Question3Choice1",
            "ui": Object {
              "title": "group1Question3Choice1",
            },
            "value": "group1Question3Choice1",
          },
          Object {
            "disabled": false,
            "id": "group1Question3Choice2",
            "ui": Object {
              "title": "group1Question3Choice2",
            },
            "value": "group1Question3Choice2",
          },
          Object {
            "disabled": false,
            "id": "group1Question3Choice3",
            "ui": Object {
              "title": "group1Question3Choice3",
            },
            "value": "group1Question3Choice3",
          },
        ],
        "currentAnswer": Array [
          "group1Question3Choice1",
          "group1Question3Choice2",
        ],
        "disabled": false,
        "error": undefined,
        "id": "group1Question3",
        "type": "multiple",
        "ui": Object {
          "title": "group1Question3",
        },
        "validatedAnswer": Array [
          "group1Question3Choice1",
          "group1Question3Choice2",
        ],
        "validating": false,
      },
    ],
    "ui": Object {},
  },
]
`;

exports[`Default Answers resetAnswer 2`] = `Object {}`;

exports[`Default Answers resetAnswer 3`] = `
Object {
  "group1Question1": "group1Question1Value",
  "group1Question2": "group1Question2Choice1",
  "group1Question3": Array [
    "group1Question3Choice1",
    "group1Question3Choice2",
  ],
  "subGroup1Question1": "subGroup1Question1Value",
  "subGroup1Question2": "subGroup1Question2Choice1",
  "subGroup1Question3": Array [
    "subGroup1Question3Choice1",
    "subGroup1Question3Choice2",
  ],
  "subGroup1Question4": Array [
    "subGroup1Question4Choice1",
    "subGroup1Question4Choice2",
  ],
}
`;

exports[`Default Answers resetAnswer 4`] = `
Object {
  "group1Question1": "group1Question1Value",
  "group1Question2": "group1Question2Choice1",
  "group1Question3": Array [
    "group1Question3Choice1",
    "group1Question3Choice2",
  ],
  "subGroup1Question1": "subGroup1Question1Value",
  "subGroup1Question2": "subGroup1Question2Choice1",
  "subGroup1Question3": Array [
    "subGroup1Question3Choice1",
    "subGroup1Question3Choice2",
  ],
  "subGroup1Question4": Array [
    "subGroup1Question4Choice1",
    "subGroup1Question4Choice2",
  ],
}
`;

exports[`Default Answers resetGroup 1`] = `
Array [
  Object {
    "disabled": false,
    "groups": Array [
      Object {
        "disabled": false,
        "groups": Array [],
        "id": "subGroup1",
        "questions": Array [
          Object {
            "choices": Array [],
            "currentAnswer": "subGroup1Question1Value",
            "disabled": false,
            "error": undefined,
            "id": "subGroup1Question1",
            "type": "any",
            "ui": Object {
              "title": "subGroup1Questions",
            },
            "validatedAnswer": "subGroup1Question1Value",
            "validating": false,
          },
          Object {
            "choices": Array [
              Object {
                "disabled": false,
                "id": "subGroup1Question2Choice1",
                "ui": Object {
                  "title": "subGroup1Question2Choice1",
                },
                "value": "subGroup1Question2Choice1",
              },
              Object {
                "disabled": false,
                "id": "subGroup1Question2Choice2",
                "ui": Object {
                  "title": "subGroup1Question2Choice2",
                },
                "value": "subGroup1Question2Choice2",
              },
              Object {
                "disabled": false,
                "id": "subGroup1Question2Choice3",
                "ui": Object {
                  "title": "subGroup1Question2Choice3",
                },
                "value": "subGroup1Question2Choice3",
              },
            ],
            "currentAnswer": "subGroup1Question2Choice1",
            "disabled": false,
            "error": undefined,
            "id": "subGroup1Question2",
            "type": "single",
            "ui": Object {
              "title": "subGroup1Question2",
            },
            "validatedAnswer": "subGroup1Question2Choice1",
            "validating": false,
          },
          Object {
            "choices": Array [
              Object {
                "disabled": false,
                "id": "subGroup1Question3Choice1",
                "ui": Object {
                  "title": "subGroup1Question3Choice1",
                },
                "value": "subGroup1Question3Choice1",
              },
              Object {
                "disabled": false,
                "id": "subGroup1Question3Choice2",
                "ui": Object {
                  "title": "subGroup1Question3Choice2",
                },
                "value": "subGroup1Question3Choice2",
              },
              Object {
                "disabled": false,
                "id": "subGroup1Question3Choice3",
                "ui": Object {
                  "title": "subGroup1Question3Choice3",
                },
                "value": "subGroup1Question3Choice3",
              },
            ],
            "currentAnswer": Array [
              "subGroup1Question3Choice1",
              "subGroup1Question3Choice2",
            ],
            "disabled": false,
            "error": undefined,
            "id": "subGroup1Question3",
            "type": "multiple",
            "ui": Object {
              "title": "subGroup1Question3",
            },
            "validatedAnswer": Array [
              "subGroup1Question3Choice1",
              "subGroup1Question3Choice2",
            ],
            "validating": false,
          },
          Object {
            "choices": Array [
              Object {
                "disabled": false,
                "id": "subGroup1Question4Choice1",
                "ui": Object {
                  "title": "subGroup1Question4Choice1",
                },
                "value": "subGroup1Question4Choice1",
              },
              Object {
                "disabled": false,
                "id": "subGroup1Question4Choice2",
                "ui": Object {
                  "title": "subGroup1Question4Choice2",
                },
                "value": "subGroup1Question4Choice2",
              },
              Object {
                "disabled": false,
                "id": "subGroup1Question4Choice3",
                "ui": Object {
                  "title": "subGroup1Question4Choice3",
                },
                "value": "subGroup1Question4Choice3",
              },
            ],
            "currentAnswer": Array [
              "subGroup1Question4Choice1",
              "subGroup1Question4Choice2",
            ],
            "disabled": false,
            "error": undefined,
            "id": "subGroup1Question4",
            "type": "multiple",
            "ui": Object {
              "title": "subGroup1Question3",
            },
            "validatedAnswer": Array [
              "subGroup1Question4Choice1",
              "subGroup1Question4Choice2",
            ],
            "validating": false,
          },
        ],
        "ui": Object {},
      },
    ],
    "id": "group1",
    "questions": Array [
      Object {
        "choices": Array [],
        "currentAnswer": "group1Question1Value",
        "disabled": false,
        "error": undefined,
        "id": "group1Question1",
        "type": "any",
        "ui": Object {
          "title": "group1Questions",
        },
        "validatedAnswer": "group1Question1Value",
        "validating": false,
      },
      Object {
        "choices": Array [
          Object {
            "disabled": false,
            "id": "group1Question2Choice1",
            "ui": Object {
              "title": "group1Question2Choice1",
            },
            "value": "group1Question2Choice1",
          },
          Object {
            "disabled": false,
            "id": "group1Question2Choice2",
            "ui": Object {
              "title": "group1Question2Choice2",
            },
            "value": "group1Question2Choice2",
          },
          Object {
            "disabled": false,
            "id": "group1Question2Choice3",
            "ui": Object {
              "title": "group1Question2Choice3",
            },
            "value": "group1Question2Choice3",
          },
        ],
        "currentAnswer": "group1Question2Choice1",
        "disabled": false,
        "error": undefined,
        "id": "group1Question2",
        "type": "single",
        "ui": Object {
          "title": "group1Question2",
        },
        "validatedAnswer": "group1Question2Choice1",
        "validating": false,
      },
      Object {
        "choices": Array [
          Object {
            "disabled": false,
            "id": "group1Question3Choice1",
            "ui": Object {
              "title": "group1Question3Choice1",
            },
            "value": "group1Question3Choice1",
          },
          Object {
            "disabled": false,
            "id": "group1Question3Choice2",
            "ui": Object {
              "title": "group1Question3Choice2",
            },
            "value": "group1Question3Choice2",
          },
          Object {
            "disabled": false,
            "id": "group1Question3Choice3",
            "ui": Object {
              "title": "group1Question3Choice3",
            },
            "value": "group1Question3Choice3",
          },
        ],
        "currentAnswer": Array [
          "group1Question3Choice1",
          "group1Question3Choice2",
        ],
        "disabled": false,
        "error": undefined,
        "id": "group1Question3",
        "type": "multiple",
        "ui": Object {
          "title": "group1Question3",
        },
        "validatedAnswer": Array [
          "group1Question3Choice1",
          "group1Question3Choice2",
        ],
        "validating": false,
      },
    ],
    "ui": Object {},
  },
]
`;

exports[`Default Answers resetGroup 2`] = `Object {}`;

exports[`Default Answers resetGroup 3`] = `
Object {
  "group1Question1": "group1Question1Value",
  "group1Question2": "group1Question2Choice1",
  "group1Question3": Array [
    "group1Question3Choice1",
    "group1Question3Choice2",
  ],
  "subGroup1Question1": "subGroup1Question1Value",
  "subGroup1Question2": "subGroup1Question2Choice1",
  "subGroup1Question3": Array [
    "subGroup1Question3Choice1",
    "subGroup1Question3Choice2",
  ],
  "subGroup1Question4": Array [
    "subGroup1Question4Choice1",
    "subGroup1Question4Choice2",
  ],
}
`;

exports[`Default Answers resetGroup 4`] = `
Object {
  "group1Question1": "group1Question1Value",
  "group1Question2": "group1Question2Choice1",
  "group1Question3": Array [
    "group1Question3Choice1",
    "group1Question3Choice2",
  ],
  "subGroup1Question1": "subGroup1Question1Value",
  "subGroup1Question2": "subGroup1Question2Choice1",
  "subGroup1Question3": Array [
    "subGroup1Question3Choice1",
    "subGroup1Question3Choice2",
  ],
  "subGroup1Question4": Array [
    "subGroup1Question4Choice1",
    "subGroup1Question4Choice2",
  ],
}
`;
